### ISSUE TITLE
Account for 'type: module' in package.json when doing ES Modules detection

### ISSUE THREAD

**Issue by michaeljones** [pausan/mocha-list-tests#7]
It seems that the code is checking if there is a '.mjs' extension in order to pick between require & import. I think that if the `package.json` specifies [`type` to be `module`](https://nodejs.org/api/packages.html#packages_type) then all `.js` files should be treated as ESM. I guess that only applies to files in the main project. It  only impacts files where the nearest `package.json` as `type` `module` so anything in `node_modules` or sub-directories with their own `package.json` files might be different. 

I'm finding the whole esm/mjs/type-module set up very confusing and I would say that no action is required for my benefit at the moment as I don't think the ecosystem as a whole has settled down enough in this manner. Typescript 4.5 has only just been released with support for generating `.mjs` files from `.mts` files so it seems like stuff is still being worked out. It seems like it might be good to support `type` `module` at some point though. I don't know if there is a library out there that helps answer the question of 'does the nearest package.json to this file specify `type` `module` or not` but I hope there is at some point!

**Comment by pausan** [comment]
Yeah, the very first time I learned about mjs and js distinction things were a little confusing, since you can mix both in different scenarios (and honestly, I'm not sure I can recall all scenarios).

Well, I can think of some approaches here:
1) Explore top directories until root to find the first package.json and then load it and look for type: module, and then treat all .js as ".mjs"
2) Try with current import/require inside a try/catch and then if it fails, catch the error and try the other way around
3) Read the file, look for "import" vs "require" patterns and load accordingly.
4) Let the caller of the functions decide what to use (allow to force one way or the other); drawback of this is that if a project has mixed mjs/js types, it might not work on some scenarios

Since this library/tool it is used at a folder level, and not at a project level, it is hard to tell if it would be using the right loader.

I believe the biggest issue in order to identify the right loader for mjs vs js is when ".js" extension is used but you have package.json saying type=module on some higher-level folder;  I would say probably option 1 would work on the scenarios I can think of, but I'm not sure if that would cover all scenarios... otherwise I would go for option 2. I get the feeling option 2, although more hacky, would cover all possible scenarios if it works, but option 1 definitely looks less hacky.

I think I won't be implementing this unless somebody provides at least one simple test folder structure reproducing the issue. I would accept any PR including a test case + the fix that I can reproduce.
